MCP.1 configuration portion

jni2000 opened the pull request:
This is a preliminary implementation of MCP.1. Work is still in progress. We would like to collect early comments and make sure that we do not by any chance go wrong too far. This pull request is not really for merge but rather to get inputs from everyone and make sure we are inline with the implementation. So please help provide your valuable comments. Thanks.

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+#    Copyright 2016 SAS Project Authors. All Rights Reserved.
```
There is no need to create a new type of testcase -- this can live in the MCP file directly.

jni2000 replied:
This   file is introduced for all MCP and related test cases, going forward we will   add all helper and utility methods in this file

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+
+  def setUp(self):
+    self._sas, self._sas_admin = sas.GetTestingSas()
+    # Note that we disable the admin.Reset(), to avoid 'unexpected' request
```
This is not needed here.

jni2000 replied:
Modified,   removed the extra initializations and methods

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+    # to SAS UUT before a test really starts.
+    # Tests changing the SAS UUT state must explicitly call the SasReset()
+    #self.SasReset()
+    self.sasTHInstances = {}
```
Please use consistent variable naming conventions.

jni2000 replied:
updated   with sas_th_harness

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+    #self.SasReset()
+    self.sasTHInstances = {}
+
+  def SasReset(self):
```
Not needed.

jni2000 replied:
updated,   removed the reset method

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+
+  def tearDown(self):
+     # check if any SASTH is started and try to stop it
+     for sasTH in sorted(self.sasTHInstances.itervalues()):
```
Does it need to be sorted?

jni2000 replied:
updated,   removed the sort

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+     for sasTH in sorted(self.sasTHInstances.itervalues()):
+         sasTH.shutdown()
+
+  def createSASTH(self,sasTHConfig):
```
Please fix the name of this method and all following method/variable names to be consistent with the established style of the WinnForum test code.

jni2000 replied:
updated   with "create_sas_th"

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+         sasTH.shutdown()
+
+  def createSASTH(self,sasTHConfig):
+     sasTHSettings = sasTHConfig['sas_hareness']
```
Typo: "harness" should be "harness".

jni2000 replied:
updated   all the instances of 'hareness' with 'harness'

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+  def createSASTH(self,sasTHConfig):
+     sasTHSettings = sasTHConfig['sas_hareness']
+     # prepare a HTTPServer Object for each sasTH mentioned in configuration file
+     for item in sasTHSettings:
```
Can you pick a more descriptive name than `item`?

jni2000 replied:
replaced   the name 'item' with 'sas'

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+       '''
+       end_time = datetime.now()
+       time_delta = timedelta(hours=-24)
+       start_time = end_time.today() + time_delta
```
It's more readable to use
`start_time = end_time.today() - timedelta(hours=-24)`

This also means you don't have to reuse variable names.

jni2000 replied:
This   method is used by SAS Test Harness, Modifications will be taken as part of   test harness enhancements

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+       end_time = datetime.now()
+       time_delta = timedelta(hours=-24)
+       start_time = end_time.today() + time_delta
+       end_time = str(end_time.replace(microsecond=0).isoformat()) + 'Z'
```
Why not use datetime.strftime() ?

jni2000 replied:
This   method is used by SAS Test Harness, Modifications will be taken as part of   test harness enhancements

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+       end_time = datetime.now()
+       time_delta = timedelta(hours=-24)
+       start_time = end_time.today() + time_delta
+       end_time = str(end_time.replace(microsecond=0).isoformat()) + 'Z'
```
Please avoid reusing variable names.

jni2000 replied:
This   method is used by SAS Test Harness, Modifications will be taken as part of   test harness enhancements

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+       grant_expire = str(grant_expire.replace(microsecond=0).isoformat()) + 'Z'
+       return (start_time,end_time,grant_expire)
+    
+     configDir = 'configs/test_WINNF_FT_S_MCP_1' 
```
Please do not statically configure this -- you can automatically read in the config using the @configurable_testcase decorator.

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+       return (start_time,end_time,grant_expire)
+    
+     configDir = 'configs/test_WINNF_FT_S_MCP_1' 
+     for sasTHName,sasTHInstance in sorted(self.sasTHInstances.items()):
```
Why sort?

jni2000 replied:
Updated,   sort is removed

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+     configDir = 'configs/test_WINNF_FT_S_MCP_1' 
+     for sasTHName,sasTHInstance in sorted(self.sasTHInstances.items()):
+        startTime,endTime,grant_expire = getTimeStamp()
+        # load dumpfile which starts with SAS-TH name to identify easily specfic CBSD dump record associated with corresponding SAS-TH
```
I don't follow, can you explain?

jni2000 replied:
This   logic is used to create the FAD records by SAS Test Harness, as mentioned all   the optimizations for SAS TH will be taken as part of SAS TH enhancements

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,125 @@
+        startTime,endTime,grant_expire = getTimeStamp()
+        # load dumpfile which starts with SAS-TH name to identify easily specfic CBSD dump record associated with corresponding SAS-TH
+        sasTHName = sasTHName + '-' + iterationNumber
+        dumpfiles = [dumpfile for dumpfile in os.listdir(configDir)
```
The dump files will be part of the config itself, not separate files in the config directory.

Each file in the config directory should be a single *complete* configuration of MCP.1.

jni2000 replied:
Updated,   all the required data will be present in the config file now

kate-harrison commented on src/harness/sas_test_harness.py:
```diff
@@ -0,0 +1,143 @@
+import ssl
```
This looks like the same code from #368 so I will skip reviewing it in this PR.

jni2000 replied:
Yes,   used PR 368 code here

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+#    Copyright 2017 SAS Project Authors. All Rights Reserved.
```
2018

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+    grant_record_4 = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_3.json')))
+    
+    #Load CBSD
```
Prefer to load CBSDs before grant requests.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+  def generate_MCP_1_default_config(self,filename):
+    """ Generates the WinnForum configuration for MCP.1. """
+    # Load Grant requests
+    grant_record_1 = json.load(
```
`grant_request_1` would be more consistent with existing naming conventions.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+                        'cert_file': 'certs/server.cert',
+                        'key_file': 'certs/server.key'}, 
+        'K':3, 
+        'headroom':{'MgPpa':2,'MgGwpz':2,'MgCochannel':2,'MgBlocking':2,'MgOobe':2,'MgEsc':2},
```
Headroom is *not* configurable.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+
+    # Create the actual config.
+    config = {
+        'cbsdRecords': [{'registrationRequests':[device_1,device_2],'grantRequests':[grant_record_1,grant_record_2]},
```
Please use a format consistent with existing config files. Please also fix the formatting so that this is more readable.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+    config = {
+        'cbsdRecords': [{'registrationRequests':[device_1,device_2],'grantRequests':[grant_record_1,grant_record_2]},
+                        {'registrationRequests':[device_3,device_4],'grantRequests':[grant_record_3,grant_record_4]}],
+        'IAPRecords': [[
```
Prefer to keep the information for a single iteration close together. So the final config might look like:
`[ iter1_config, iter2_config, ...]`

That way it's easier to tell what's going on at each step.

jni2000 replied:
updated,   added iteration based details to 'cbsdRecords' and 'IAP Records'

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+        'cbsdRecords': [{'registrationRequests':[device_1,device_2],'grantRequests':[grant_record_1,grant_record_2]},
+                        {'registrationRequests':[device_3,device_4],'grantRequests':[grant_record_3,grant_record_4]}],
+        'IAPRecords': [[
+      { 'type': 'FSS','Record': [fss_record_1] },
```
Could you have a dictionary whose key is the type name and the value is the list of records?

jni2000 replied:
Updated   the IAP records, dictionary with type and values as list of records

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+			'cert_file': 'certs/server_b.cert',
+			'key_file': 'certs/server_b.key'
+		        }],
+        'sas_uut': {'name': 'SAS-UUT',
```
This is already in sas.cfg.

jni2000 replied:
updated,   removed the sas_uut info from config

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+	'sas_hareness': [{'name': 'SAS-TH-1',
+			'url': ['localhost', '9001'],
+			'cert_file': 'certs/server_a.cert',
+			'key_file': 'certs/server_a.key'},
```
The FAD contents should be in the config as well.

jni2000 replied:
Udates,   included the fad records in the conig along with 'sas_harness' info

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+                        'key_file': 'certs/server.key'}, 
+        'K':3, 
+        'headroom':{'MgPpa':2,'MgGwpz':2,'MgCochannel':2,'MgBlocking':2,'MgOobe':2,'MgEsc':2},
+        'DPAactiveInfo':[],
```
Prefer to include an example of this (but OK to leave as a TODO for now since there is already a lot to be done).

jni2000 replied:
Updated   the dpa active and deactive list

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+    """SAS manages a mix of GAA and PAL Grants in 3550 MHz
+       to 3700 MHz to protect configurable IAP-protected entities and DPAs
+    """
+    config = loadConfig(config_filename)
```
Strongly suggest that we do some light checking of the config.

jni2000 replied:
Added,   light check on config

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       to 3700 MHz to protect configurable IAP-protected entities and DPAs
+    """
+    config = loadConfig(config_filename)
+    pre_iap_headroom = config ['headroom']
```
Let's have this information stored in a file similar to sas.cfg -- the headroom value will be common to all SASes and does not need to be configurable within the test case.

jni2000 replied:
we will   check if any other MCP related test cases have dependency? For all the MCP   related TCs we will add a new configuration file

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+    self.createSASTH(config)
+    self.configureFadForEachSASTH(0)
+    
+    #fad objects get the fad dumps from the remote SAS configured with 'url'
```
It looks like this is actually fetching from the local SAS Test Harnesses, no?

jni2000 replied:
here we   are configuring the SAS Test harness with FAD objects as per the step 3 in   test case

kate-harrison commented on src/harness/testcases/configs/test_WINNF_FT_S_MCP_1/sas_hareness.config:
```diff
@@ -0,0 +1,16 @@
+{
```
Is this used?

jni2000 replied:
updated,   included in the default.config

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       fad_object_th.append(FullActivityDump(cert,key,config['sas_hareness'][N]['url']))
+       fad_records.append(fad_object_th[N])
+
+    #for UUT creating fad object
```
Please note that the test spec requires that *each* SAS TH pull from the SAS UUT.

jni2000 replied:
Require   changes in SAS Test Harness implementation. We will address this comment as   part of SAS TH development activity

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+
+    # informing SAS UUT about SAS TH 
+    for N in len(config['sas_hareness']):
+       self._sas_admin.InjectPeerSas({'certificateHash': config[N]['certificateHash'],
```
Cert hash should be automatically computed from the cert that's provided for the SAS TH to minimize errors.

jni2000 replied:
Updated,   referred the PR368

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+    # informing SAS UUT about SAS TH 
+    for N in len(config['sas_hareness']):
+       self._sas_admin.InjectPeerSas({'certificateHash': config[N]['certificateHash'],
+                                     'url': 'https://' + config[N] ['url']+ '/dump'}) 
```
Do not include the "/dump" at the end.

jni2000 replied:
Referred   this API from PR368

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+    
+
+    #Step 4,5 : Inject IAP protected entities into UUT
+    for k in  config['K']:	
```
This does not work since config['K'] is a scalar.

jni2000 replied:
updated   with range(config['K'])

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+             self._sas_admin.InjectEsc(records['Record'])
+          if records['type'] == "PPA":
+             self._sas_admin.InjectZoneData(records['Record'])
+
```
This should be if-elif-elif-else where the 'else' gives a failure (to detect config problems).

jni2000 replied:
Updated,   modified the logic of Invoking the API

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       if len(config['sas_hareness']) > 0 :
+          self._sas_admin.TriggerFullActivityDump() # triggers to UUT
+
+          # Step 7 : Pull FAD records from SAS UUT
```
Please check the test spec -- each SAS TH has to pull from SAS UUT.

jni2000 replied:
Require   changes in SAS Test Harness implementation. We will address this comment as   part of SAS TH development activity

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+
+       # Step 9 : Call IAP reference model, calculated at step 12
+       #collect all the protected Entites
+       protectedEntities.append(config['IAPRecords'][k])
```
Has this variable been defined yet?

jni2000 replied:
Updated,   defined the variable

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+                fad_records_iap = fad_records
+          #Call IAP reference model, fad_records contains fad_UUT and fad_SAS_TH
+          #To-Do : If there is any change in the API defination, parameters need to be updated
+          iap_refmodel_margin_list = IapReferenceModel(config['headroom'], pe,fad_records_iap)
```
*All* of the pre-IAP filtering needs to be done before you can run the IAP reference model for any protected entity.

jni2000 replied:
Added,   Invoked the pre-IAP API

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       # Step 10 : Register N(2,k)CBSDs with SAS UUT
+       # Use DP objects to get CBSD registered
+       # creation of Domain Proxy Object
+       dp_cert = os.path.join('certs', 'dp_client.cert')
```
Should be configurable.

jni2000 replied:
Updated,   certificated are configured now

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       # creation of Domain Proxy Object
+       dp_cert = os.path.join('certs', 'dp_client.cert')
+       dp_key = os.path.join('certs', 'dp_client.key')
+       reg_records = config['cbsdRecords'][k]['registrationRequests']
```
`registration_requests` -- please use consistent terminology.

jni2000 replied:
Updated   to registration_requests

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       reg_records = config['cbsdRecords'][k]['registrationRequests']
+       grant_records = config['cbsdRecords'][k]['grantRequests']	
+       #dp.initialize triggers the registration and grant for the records passes as parameters
+       dp = DomainProxy(dp_cert,dp_key,reg_records,grant_records)
```
The DP needs to persist through iterations, not be created and destroyed each time. Otherwise the test harness will forget about CBSDs that the SAS still knows about.

jni2000 replied:
Updated,   invoked dp.Initialize method

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+                fad_records_iap = fad_records
+          #Call IAP reference model, fad_records contains fad_UUT and fad_SAS_TH
+          #To-Do : If there is any change in the API defination, parameters need to be updated
+          iap_refmodel_margin_list = IapReferenceModel(config['headroom'], pe,fad_records_iap)
```
What happens to this var? It seems like it will not contain information about all protected entities, which is what we need for the CHECK step.

jni2000 replied:
Updated,   As per the IAP reference document

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+          agg_interference_iap_list = AggregateInterferenceModelWithIAP(cbsd_agg_list,pe)
+          
+          #check MCP.1
+          checkMCP_IAP(pe,agg_interference_list,agg_interference_iap_list,pre_iap_headroom,leftOverAlloc)
```
`leftoverAlloc` hasn't been initialized.

Also the output of the IAP ref model should already include the headroom and leftover allocations, so there's no need to include them here.

Finally, we need to include `delta_IAP` which should be configurable.

jni2000 replied:
Updated   the API as per the IAP reference model API definition

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+ 
+       # Step 18,19,20 and 21 : 
+       # Send heartbeat request for the grants, relinquish the grant, grant request and heartbeat for new grant
+       dp.HeartBeatRequestForAllGrants_relinquish()
```
I don't see the implementation of this method in PR #404.

jni2000 replied:
Updated,   aligned with PR404 methods

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       #step 24: wait for 240 sec if DPA is activated in step 23 else 15 sec
+       if config['DPAactiveInfo'][k] :
+          TriggerDpaActivation(config['DPAactiveInfo'][k])
+          sleep (240)
```
This is not defined.

jni2000 replied:
Added,   imported the time module

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       for pe in protectedEntities :
+ 
+          #invoking Aggregate Interference API,
+          agg_interface_list = AggregateInterferenceModel(cbsd_agg_list,pe)
```
`interface` should be `interference`

jni2000 replied:
updated,   interface with interference

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+  def checkMCP_IAP(self,pe,agg_interface_list,iap_refmodel_margin_list,pre_iap_headroom,left_over_alloc):
+    """ Performs check step """
+    # To Do
+    if ((pe['type'] == 'FSS') or (pe['type'] == 'ESC')): # only one lelement with fifferent frequency
```
Please fix typos.

jni2000 replied:
Updated,   modified the logic of MCP check

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       for p in len(agg_interface_list):
+          if (agg_interface_list[p] <= (iap_refmodel_margin_list[p] +pre_iap_headroom + leftOverAlloc)):
+             points_valid = points_valid +1
+          else:
```
`else: continue` is superfluous.

jni2000 replied:
Updated,   modified the logic of MCP check

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+             points_valid = points_valid +1
+          else:
+             continue
+       self.assertLessEqual(points_valid , (p*0.95))
```
`p` is not a number.

jni2000 replied:
Updated,   logic is modified

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       self.assertLessEqual(points_valid , (p*0.95))
+
+  def removePurgeCbsd(fad_records,cbsd_purge_list):
+    """ Remove purged cbsd """
```
Implementation missing?

jni2000 replied:
This   Functionality is implemented as part of pre-IAP modules

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,328 @@
+       # Step 16 : Trigger CPAS and wait for its completion
+       self.self.TriggerDailyActivitiesImmediatelyAndWaitUntilComplete()
+       
+       # Step 17 : Call IAP reference model
```
Since this occurs in steps 9 and 17, it's probably best to abstract it into a function to ensure consistency.

jni2000 replied:
Updated,   Invoked pre-IAP module

jni2000 commented:
Per our offline discussion and agreement, here we submit the revised preliminary MCP.1 code focusing mainly on the configuration part of it to allow early review. If the configuration is reviewed and approved, we will be able to move on and customize and submit all the related xPR test cases. Please help review the configuration part of this test case and give comments. We can address other parts later.

Also, we will upload our responses to the previous review comments soon.

Thanks,

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+                         'MgBlocking': 2, 'MgOobe': 2, 'MgEsc': 2},
+            'dpaActiveList': [dpa_0],
+            'dpaInactiveList': [dpa_1],
+            'protectionThreshold': {'QPpa': -80, 'QGwpz': -80,
```
This should be part of the overall configuration of the test harness, and not the configuration for a particular test case.

jni2000 replied:
Removed   from testcase file, added new file " sas_test_harness.cfg" for SAS   Test Harness configurations.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+            'sasHarness': [sas_harness_0, sas_harness_1],
+            'dpConfig': {'cert': 'dp_client.cert', 'key': 'dp_client.key'},
+            'K': 2,
+            'headroom': {'MgPpa': 2, 'MgGwpz': 2, 'MgCochannel': 2,
```
As mentioned in a previous review, headroom is *not* configurable for a given test case, it will instead be part of the overall configuration of the test harness. Please do not include it in the config file for MCP.1.

jni2000 replied:
Removed   from testcase file, added new file " sas_test_harness.cfg" for SAS   Test Harness configurations.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+                }
+        # Create the actual config.
+        config = {
+            'cbsdRecordsN1': [cbsd_records_iter_1, cbsd_records_iter_2],
```
As mentioned previously...

Prefer to keep the information for a single iteration close together. So the final config might look like:

`[ iter1_config, iter2_config, ...]`

That way it's easier to tell what's going on at each step.

If my suggestion is confusing, please let me know and I will try to add more detail.

jni2000 replied:
Updated,   modified the configurations as suggested

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+                }
+        # Create the actual config.
+        config = {
+            'cbsdRecordsN1': [cbsd_records_iter_1, cbsd_records_iter_2],
```
No need to use "N1" here.

jni2000 replied:
Updated,   Removed N1

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+        # Create the actual config.
+        config = {
+            'cbsdRecordsN1': [cbsd_records_iter_1, cbsd_records_iter_2],
+            'protectedEntitiesPIAP': [protected_entities_iter_1, protected_entities_iter_2],
```
This could be simply "protectedEntities".

jni2000 replied:
Updated,   Changed the name to "protectedEntities"

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+            'cbsdRecordsN1': [cbsd_records_iter_1, cbsd_records_iter_2],
+            'protectedEntitiesPIAP': [protected_entities_iter_1, protected_entities_iter_2],
+            'sasHarness': [sas_harness_0, sas_harness_1],
+            'dpConfig': {'cert': 'dp_client.cert', 'key': 'dp_client.key'},
```
The test case needs to support multiple DPs.

jni2000 replied:
Updated   , Multiple DPs support Is enabled.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+            'protectedEntitiesPIAP': [protected_entities_iter_1, protected_entities_iter_2],
+            'sasHarness': [sas_harness_0, sas_harness_1],
+            'dpConfig': {'cert': 'dp_client.cert', 'key': 'dp_client.key'},
+            'K': 2,
```
The value of K should be implicit in the configuration if you follow the structure I suggested previously. This will reduce the opportunity for self-inconsistent configs (in fact, the current config contradicts itself as you earlier assumed that you will have at least 3 iterations).

jni2000 replied:
Updated   , Removed 'K'

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+            'K': 2,
+            'headroom': {'MgPpa': 2, 'MgGwpz': 2, 'MgCochannel': 2,
+                         'MgBlocking': 2, 'MgOobe': 2, 'MgEsc': 2},
+            'dpaActiveList': [dpa_0],
```
We need to be able to activate/deactivate DPAs at each iteration, not just once. Please note also that a DPA can only be deactivated if it has previously been activated, since the test case begins with all DPAs in the inactive state.

jni2000 replied:
Updated,   configured with multiple DPA records

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+        sas_harness_0 = {
+                    'name': 'SAS-TH-1',
+                    'url': 'localhost:9001',
+                    'cert_file': 'certs/server_a.cert',
```
You seem to assume two different conventions for specifying the cert and key: one prepends with "certs/" and the other gives only the filename. I'm not sure which is better (probably the latter?) but we should at least be consistent.

jni2000 replied:
Updated   , Used latter format

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+                    'url': 'localhost:9001',
+                    'cert_file': 'certs/server_a.cert',
+                    'key_file': 'certs/server_a.key',
+                    'certificateHash': 'dummy',
```
This should be a correct value computed from the actual cert. Leave a comment if you intend to update before submitting.

jni2000 replied:
This   configuration is for SAS TH. Parameters are configured based on the inputs   received from the SAS TH development. Which is is under #PR368, we will   update the SAS TH configurations section after #PR386 merge

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+                    'key_file': 'certs/server_a.key',
+                    'certificateHash': 'dummy',
+                    'cbsdRecord': [sas_th_0_cbsd_iter_0, sas_th_0_cbsd_iter_1, sas_th_0_cbsd_iter_2],
+                    'fileServePath': 'sas_harness_2/'
```
Does this need to be configured? Probably it could be randomly generated in the test code without causing any trouble. That way we reduce the chance for the tester to introduce a copy-and-paste error which has two SAS THs using the same data directory.

jni2000 replied:
This   configuration is for SAS TH. Parameters are configured based on the inputs   received from the SAS TH development. Which is is under #PR368, we will   update the SAS TH configurations section after #PR386 merge

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+                              'grantRequests': [grant_request_1, grant_request_2]}
+        cbsd_records_iter_2 = {'registrationRequests': [device_3, device_4],
+                              'grantRequests': [grant_request_3, grant_request_4]}
+        protected_entities_iter_1 = {'PPA': [ppa_record_1], 'ESC': [esc_record_1]}
```
In order to inject a PPA, we also have to inject the corresponding PAL records.

jni2000 replied:
Updated by adding PalAndPpa field in protectedEntities which has a set of   pal and ppa records

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+        sas_harness_0 = {
+                    'name': 'SAS-TH-1',
+                    'url': 'localhost:9001',
+                    'cert_file': 'certs/server_a.cert',
```
Please use `camelCase` rather than `snake_case` for variable names, in order to be consistent with the rest of the test code.

jni2000 replied:
Updated,   modified as per the comments

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+        ppa_record_1 = json.load(
+          open(os.path.join('testcases', 'testdata', 'ppa_record_0.json'))) 
+        # Registration configuration for iteration 1
+        cbsd_records_iter_1 = {'registrationRequests': [device_1, device_2],
```
You will need to include the ability to inject conditional registration information so that you can use Cat B CBSDs as well.

jni2000 replied:
Updated, by adding conditional data information as part of registration   data

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+          open(os.path.join('testcases', 'testdata', 'device_c.json')))
+        device_4 = json.load(
+          open(os.path.join('testcases', 'testdata', 'device_d.json')))
+        device_5 = json.load(
```
This appears to be unused.

jni2000 replied:
Updated,   Removed now

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+        sas_th_0_cbsd_iter_1 = [sas_th_cbsd_1]
+        sas_th_0_cbsd_iter_2 = [sas_th_cbsd_2]
+        sas_th_1_cbsd_iter_0 = [sas_th_cbsd_3]
+        sas_th_1_cbsd_iter_1 = [sas_th_cbsd_4]
```
Building up the data inside the SAS TH, instead of constantly swapping it out, would probably be a better test of the system. It would be ideal if the default config looked more like what we believe to be a reasonable test case.

jni2000 replied:
This   configuration is for SAS TH. Parameters are configured based on the inputs   received from the SAS TH development. Which is is under #PR368, we will   update the SAS TH configurations section after #PR386 merge

kate-harrison replied:
Good point. I will review the SAS TH configuration in #368 instead, and focus the comments for this PR on the issues which are specific to MCP. I suspect we will submit that PR first anyway so it should not cause any problems.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,378 @@
+        config = {
+            'cbsdRecordsN1': [cbsd_records_iter_1, cbsd_records_iter_2],
+            'protectedEntitiesPIAP': [protected_entities_iter_1, protected_entities_iter_2],
+            'sasHarness': [sas_harness_0, sas_harness_1],
```
The name should reflect the fact that multiple SAS Test Harness configs are provided.

jni2000 replied:
Updated   accordingly,  made it sas_harnesses

jni2000 commented:
Kate, we have reacted to your review comments and did corresponding adjustments and corrections. Please move on to check if this configuration part is good for moving on. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+        cbsd_records_iter_2 = {'registrationRequests': [device_3, device_4],
+                               'grantRequests': [grant_request_3, grant_request_4]}
+        pal_and_ppa_record_1 = {'PAL': pal_record_1, 'PPA': ppa_record_1}
+        protected_entities_iter_1 = {'PalAndPpa': [pal_and_ppa_record_1], 'ESC': [esc_record_1]}
```
I think this structure makes the assumption that there is a 1-to-1 correspondence between PAL records and PPAs. In fact, one PPA can be made up of multiple PAL records. Also, a single PAL record can be used to make multiple PPAs. I propose the following structure:

`{
  "palRecords": [pal_record_1],
  "ppas": [ppa_record_1],
  "escs": [esc_record_1],
}
`

Note also the naming style, which matches established conventions.

jni2000 replied:
Updated,   as per the comment suggested.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+            'protectedEntities': protected_entities_iter_1,
+            'dpaActiveList': [dpa_1, dpa_2],
+            'dpaDeactiveList': [],
+            'deltaIap': 2
```
This is constant across iterations.

jni2000 replied:
Moved to   "config" from "iteration_config"

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+        iter1_config = {
+            'cbsdRecords': cbsd_records_iter_1,
+            'protectedEntities': protected_entities_iter_1,
+            'dpaActiveList': [dpa_1, dpa_2],
```
Prefer `dpaActivationList` to more closely match with the behavior.

jni2000 replied:
Modifed   "dpaActiveList" to "dpaActivationList "

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+                    'fileServePath': 'sas_harness_1/'
+                }
+        # Create the actual config.
+        iter1_config = {
```
This config also needs to contain the SAS TH configs (CBSD, PPA, and ESC records) for this iteration. The general SAS TH config (URL, etc.) will be discussed in #368 but I think this PR needs to determine how to specify the per-iteration data that each SAS TH will contain.

jni2000 replied:
updated   the sas TH data  with ppa and esc   records

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+
+        config = {
+            'iterationData': [iter1_config, iter2_config],
+            'sasHarnesses': [sas_harness_0, sas_harness_1],
```
`sasTestHarnesses` or `sasTestHarnessConfigs`

jni2000 replied:
Updated   to "sasTestHarnessConfigs"

kate-harrison replied:
The code does not seem to have been updated.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+                }
+        # Create the actual config.
+        iter1_config = {
+            'cbsdRecords': cbsd_records_iter_1,
```
You need a way to specify which requests will go through which DP. I suggest using a list.

jni2000 replied:
Created   list for 'cbsdRequests', each element is used corresponding domain proxy

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+                }
+        # Create the actual config.
+        iter1_config = {
+            'cbsdRecords': cbsd_records_iter_1,
```
Should this be `cbsdRequests` instead? They are not yet records.

jni2000 replied:
Modified   to " cbsdRequests"

jni2000 commented on src/harness/sas_test_harness.cfg:
```diff
@@ -0,0 +1,14 @@
+Headroom: 
+   MgPpa: 2
+   MgGwpz: 2
+   MgCochannel: 2
```
Updated   the comment with "MgFssCoChannel"

jni2000 commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+        ppa_record_1 = json.load(
+          open(os.path.join('testcases', 'testdata', 'ppa_record_0.json'))) 
+        pal_record_1 = json.load(
+          open(os.path.join('testcases', 'testdata', 'pal_record_0.json')))
```
Invoked   "makePpaAndPalRecordsConsistent" with ppa and pal records

jni2000 commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,188 @@
+        # Registration configuration for iteration 1
+        cbsd_records_iter_1 = {'registrationRequests': [device_1, device_2],
+                               'grantRequests': [grant_request_1, grant_request_2],
+                               'conditionalRequests': conditionals_iter_1
```
Modified   to ConditionalRegistrationData

jni2000 commented:
Kate and all other reviewers

We have updated our implementation on the MCP.1 configuration in response to earlier review comments from Kate et. al. Please continue review to unblock of the implementation of all related incumbent protection test cases. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -153,30 +176,29 @@ def generate_MCP_1_default_config(self, filename):
             'iterationData': [iter1_config, iter2_config],
-            'sasHarnesses': [sas_harness_0, sas_harness_1],
+            'sasHarnessesConfigs': [sas_harness_0, sas_harness_1],
             'domainProxyConfig': [{'cert': 'dp_1_client.cert', 'key': 'dp_1_client.key'},
```
Please format these lines so that they're more readable (each key/value pair on its own line, as you are doing elsewhere).

jni2000 replied:
Modified   the line format

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -153,30 +176,29 @@ def generate_MCP_1_default_config(self, filename):
             'iterationData': [iter1_config, iter2_config],
-            'sasHarnesses': [sas_harness_0, sas_harness_1],
+            'sasHarnessesConfigs': [sas_harness_0, sas_harness_1],
             'domainProxyConfig': [{'cert': 'dp_1_client.cert', 'key': 'dp_1_client.key'},
```
The variable name should be plural to indicate that it's a list of DP configs.

jni2000 replied:
Modified   to domainProxyConfigs

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -105,16 +105,28 @@ def generate_MCP_1_default_config(self, filename):
+        # Registration and grant records for multiple iterations
+        cbsd_records_iter_1_domain_proxy_0 = {'registrationRequests': [device_1, device_2],
+                                              'grantRequests': [grant_request_1, grant_request_2],
+                                              'conditionalRegistrationData': conditionals_iter_1
```
Conditional registration data is injected, it does not go through a DP.

jni2000 replied:
Agreed,   conditional data is injected through Admin interface, we kept this info along   with DP, so that we can easily pass the info to domain proxy object, which   will in turn take care of injecting the conditional registration data

kate-harrison replied:
I don't think the domain proxy object can inject the conditional registration data since this is done through the admin interface.

jni2000 replied:
Sure,   will ensure that the conditional registration data  will be injected through admin interface   before passing the registration data to domain proxy object in the test case

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -105,16 +105,28 @@ def generate_MCP_1_default_config(self, filename):
+        cbsd_records_iter_1_domain_proxy_1 = {'registrationRequests': [device_3, device_4],
+                                              'grantRequests': [grant_request_3, grant_request_4]}
+ 
+        cbsd_records_iter_2_domain_proxy_0 = {'registrationRequests': [device_1, device_2],
```
Why are you re-registering the device in the second iteration? I don't think this is allowed by the test spec.

jni2000 replied:
Updated,   new devices are used for the second iteration

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -131,20 +143,31 @@ def generate_MCP_1_default_config(self, filename):
                     'keyFile': 'server_a.key',
                     'certificateHash': 'will be provided in future',
-                    'cbsdRecord': [sas_th_0_cbsd_iter_0, sas_th_0_cbsd_iter_1, sas_th_0_cbsd_iter_2],
+                    'sasThIterationData': [sas_th_0_iteration_0_data,sas_th_0_iteration_1_data],
```
I'm concerned about the per-iteration data being far from the other per-iteration data, and I don't see a clear benefit to having it here.

jni2000 replied:
This   part of the config information is used by SAS Test Harness, so we want to   keep all the SAS TH required data in one structure at the same time want to   differentiate data records per iteration based.

kate-harrison replied:
I think it will be much easier to read and debug if we keep it with the other per-iteration data.

jni2000 replied:
updated,   now placed along with other iteration data.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -153,30 +176,29 @@ def generate_MCP_1_default_config(self, filename):
-            'dpaActiveList': [dpa_1, dpa_2],
-            'dpaDeactiveList': [],
-            'deltaIap': 2
+            'dpaActivationList': [dpa_1, dpa_2],
```
FYI (no action needed at this time) we may need to load additional information on the DPAs which are used in this test case. I'll make sure you're on the email thread for this topic so that you have the latest information.

jni2000 replied:
Thanks. 

Informational   Comment, no response required

jni2000 commented:
Thanks, Kate, we will take a look and get back to you asap.

James Ni
Director of SAS System
www.federatedwireless.com<http://www.federatedwireless.com/>
https://www.linkedin.com/pub/james-ni/4/3b9/24a
978.996.7480
[Description: Macintosh HD:Users:Sam:Dropbox (Bluetext):Federated Wireless:Creative:Logo:Logo Lock Up:Final-Lockup:Federated Wireless No Tag:FedWireless-lockup-notag-web.png]

From: Kate [mailto:notifications@github.com]
Sent: Wednesday, February 21, 2018 8:10 PM
To: Wireless-Innovation-Forum/Spectrum-Access-System
Cc: James Ni; Author
Subject: Re: [Wireless-Innovation-Forum/Spectrum-Access-System] preliminary implementation of MCP.1, still in progress, would like to… (#405)


@kate-harrison commented on this pull request.

________________________________

In src/harness/testcases/WINNF_FT_S_MCP_testcase.py<https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/405#discussion_r169829322>:

> +            'protectedEntities': protected_entities_iter_1,

+            'dpaActiveList': [dpa_1, dpa_2],

+            'dpaDeactiveList': [],

+            'deltaIap': 2

+           }

+        iter2_config = {

+            'cbsdRecords': cbsd_records_iter_2,

+            'protectedEntities': protected_entities_iter_2,

+            'dpaActiveList': [dpa_3],

+            'dpaDeactiveList': [dpa_1],

+            'deltaIap': 2

+           }

+

+        config = {

+            'iterationData': [iter1_config, iter2_config],

+            'sasHarnesses': [sas_harness_0, sas_harness_1],

The code does not seem to have been updated.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/405#discussion_r169829322>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKpZrIWukbDlgcyM2i2G-ZHlxiqErxJqks5tXL59gaJpZM4RtT2p>.

jni2000 commented:
Kate

We have responded your most recent review comments on the MCP.1 configuration part. Please take a look and let us know if this is good for us to move on customizing other related xPRs.

Thanks

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -196,8 +203,10 @@ def generate_MCP_1_default_config(self, filename):
         config = {
             'iterationData': [iter1_config, iter2_config],
             'sasHarnessesConfigs': [sas_harness_0, sas_harness_1],
```
`sasTestHarnessConfigs`

jni2000 replied:
Modified,   'sasHarnessesConfigs' to 'sasTestHarnessesConfigs'

jni2000 commented:
Kate, we are back on this MCP.1 and responded to your latest comments. Please take a look. Again, at this stage we are focusing on the configuration portion of MCP.1 now to allow early review and comments. When the configuration portion is agreed on, we can do all related xPRs in parallel. So please let us know when the configuration portion is good to go. Thanks.

kate-harrison commented on src/harness/constraint_testcase.py:
```diff
@@ -0,0 +1,110 @@
+#    Copyright 2018 SAS Project Authors. All Rights Reserved.
```
Can we remove this file from this PR? It seems to have implementation details and I'd prefer to focus on the config format for now.

jni2000 replied:
Yes. We   have removed this file from the PR.

jcereced replied:
has this file been removed (constraint_testcase.py)? i can still see it in the PR

kate-harrison commented on src/harness/testcases/testdata/cbsd_0.json:
```diff
@@ -0,0 +1,43 @@
+{
```
I don't think this (and similar files) is needed anymore.

jni2000 replied:
Yes.   This file and the similar files have been removed from the PR.

jcereced replied:
I can still see this cbsd.json files please remove

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+    pal_low_frequency = 3550000000.0
+    pal_high_frequency = 3560000000.0        
+
+    ppa_record_1, pal_record_1 = makePpaAndPalRecordsConsistent(ppa_record,
```
Please use a plural variable name for the PAL record output since it is a list.

jni2000 replied:
Modified   to ppa_records_1 and pal_records_1

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+    cbsd_records_iter_1_domain_proxy_0 = {
+        'registrationRequests': [device_1, device_2],
+        'grantRequests': [grant_request_1, grant_request_2],
+        'conditionalRegistrationData': conditionals
```
As mentioned in a previous comment, conditional registration data does not go through the DP so it does not belong here. We inject this kind of data through the admin interface, and I think it's safe to inject all data at the beginning of the test (i.e. there is no need for per-iteration reg conditional data).

jni2000 replied:
Separated   conditionals from per-iteration data to load it through admin interface.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+    
+    # Protected entities records for multiple iterations
+    protected_entities_iter_1 = {
+        'palRecords': [pal_record_1],
```
`pal_record_1` is already a list so here you would see a list of lists, which is not correct.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+
+    # SAS Test Harnesses configurations, 
+    # Following configurations are for two SAS test harnesses for two iterations
+    sas_test_harness_device_1 = json.load(
```
You need to be careful to not reuse the same (FCC ID, SN) combinations -- otherwise the SAS will reject (this logic is tested in GRA.5,6, which you are also writing).

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+      open(os.path.join('testcases', 'testdata', 'device_f.json')))
+
+    # Generate Cbsd FAD Records for SAS Test Harness 0, iteration 0
+    cbsd_fad_records_sas_test_harness_0_iteration_0 = generateCbsdRecords([sas_test_harness_device_1], [[grant_request_1]])
```
Recommend you unify your naming convention here with the one used for DPs above: `cbsd_records_iter_1_domain_proxy_0` uses iter # then DP #, whereas here you use TH # then iter #.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+        'sasTestHarnessName': 'SAS-TH-1',
+        'hostName': 'localhost',
+        'port': 9001,
+        'serverCert': "certs/server.cert",
```
Use os.path.join().

jni2000 replied:
Used   os.path.join for 'serverCert' , 'serverKey' and 'caCert'.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+        'protectedEntities': protected_entities_iter_1,
+        'dpaActivationList': [dpa_1, dpa_2],
+        'dpaDeactivationList': [],
+        'sasTestHarnessesData': [sas_test_harness_0_dump_records_iteration_0,sas_test_harness_1_dump_records_iteration_0]
```
Just `Harness`, I think? Normally you would make 'data' plural, but unfortunately in English people tend to use 'data' to refer to both singular and plural (although technically the singular should be 'datum', but that's irrelevant here :) ).

jni2000 replied:
I  agree. Modified to "sasTestHarnessData"

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+    }
+    config = {
+        'iterationData': [iteration1_config, iteration2_config],
+        'sasTestHarnesses': [sas_test_harness_0_config, sas_test_harness_1_config],
```
`sasTestHarnessConfigs` to match with FAD.2.

jni2000 replied:
Modified   to "sasTestHarnessConfigs"

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+    config = {
+        'iterationData': [iteration1_config, iteration2_config],
+        'sasTestHarnesses': [sas_test_harness_0_config, sas_test_harness_1_config],
+        'domainProxyConfigs': [{'cert': 'dp_1_client.cert', 
```
You probably need to go into the 'certs' folder to find these, right?

jni2000 replied:
Yes you   are correct. Fixed.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,243 @@
+                                                                pal_high_frequency,
+                                                                'test_user_1')
+    # Load DPA record
+    dpa_1 = json.load(
```
The format in these files does not match the format specified by the testing API. Please revise.

jni2000 replied:
Modified   as per testing API doc.

jni2000 commented:
Kate, we have done a new revision according to your latest review comments. Please take a look again and let us know if this is good for us to start the other specific protection test cases.

Also for the following two comments, I can not find the original places. I'll put them here for you to check.

You need   to be careful to not reuse the same (FCC ID, SN) combinations -- otherwise   the SAS will reject (this logic is tested in GRA.5,6, which you are also   writing). | src/harness/testcases/WINNF_FT_S_MCP_testcase.py | Done | Modified the   fccId's and userId's to be unique for all the devices loaded.
-- | -- | -- | --
Recommend   you unify your naming convention here with the one used for DPs above: cbsd_records_iter_1_domain_proxy_0 uses   iter # then DP #, whereas here you use TH # then iter #. | src/harness/testcases/WINNF_FT_S_MCP_testcase.py | Done | Modified   to use the convention followed for DPs. First the iteration# then the TH#.

jni2000 commented:
Please can we wrap up the review of the configuration part of MCP.1 as soon as we can to unblock related test cases including the FPRs that CommScope signed up to work on? I think we are very close. Please let's move this forward. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -102,48 +101,55 @@ def generate_MCP_1_default_config(self, filename):
     pal_high_frequency = 3560000000.0        
 
-    ppa_record_1, pal_record_1 = makePpaAndPalRecordsConsistent(ppa_record,
+    ppa_records_1, pal_records_1 = makePpaAndPalRecordsConsistent(ppa_record,
```
There is only one PPA record returned from this function.

jni2000 replied:
Modified   to ppa_record_1

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -102,48 +101,55 @@ def generate_MCP_1_default_config(self, filename):
                                                                 pal_low_frequency,
                                                                 pal_high_frequency,
                                                                 'test_user_1')
     # Load DPA record
```
This comment is out of date.

jni2000 replied:
Changed   the comment to "Define DPA's"

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -102,48 +101,55 @@ def generate_MCP_1_default_config(self, filename):
-    dpa_3 = json.load(
-      open(os.path.join('testcases', 'testdata', 'dpa_0.json')))
+    dpa_1 = {
+       'dpaId': 'test_dpa_id1',
```
These should be real DPA IDs.

jni2000 replied:
Modified   the DPA Ids per NTIA definition.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -102,48 +101,55 @@ def generate_MCP_1_default_config(self, filename):
+    }
+
+    # Define conditionals
+    conditionals = {
```
This var was already defined on line 76, and this new definition is invalid.

jni2000 replied:
Yes you   are right.Removed this line

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -152,85 +158,103 @@ def generate_MCP_1_default_config(self, filename):
     }
 
     # Generate SAS Test Harnesses dump records for multiple iterations   
     sas_test_harness_0_dump_records_iteration_0 = {
```
Please unify naming here as well.

jni2000 replied:
Unified   the naming conventions at appropriate places

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -152,85 +158,103 @@ def generate_MCP_1_default_config(self, filename):
     }
 
     # Create the actual config.
     iteration1_config = {
```
In some places (e.g. `sas_test_harness_0_dump_records_iteration_0` and `sas_test_harness_0_dump_records_iteration_1`) you number the iterations as 0 and 1. In other places (`iteration1_config`, `cbsd_records_iteration_1_domain_proxy_0`) you number them as 1 and 2.

Please consistently zero-index.

jni2000 replied:
Unified   the naming conventions at appropriate places

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,267 @@
+        'dpaDeactivationList': [dpa_1],
+        'sasTestHarnessData': [sas_test_harness_0_dump_records_iteration_1,sas_test_harness_1_dump_records_iteration_1]
+    }
+    config = {
```
I just realized that there's another piece missing: we need the capability for individual CBSDs (which will use their own cert/key pair) to register, get a grant, etc. So in the same way we support multiple DPs, we need to support multiple individual CBSD connections.

Sorry for catching this only at the last minute -- I literally remembered only as I was writing a reply to an unrelated email.

jni2000 replied:
We are working on this in next drop. Will update accordingly.

kate-harrison replied:
I believe that the intention is to use individual CBSDs in every step where "DP test harness" is mentioned: so ITS can use N DPs and M individual CBSDs in each of these steps. The test spec is unfortunately not very clear on this point; you are welcome to double-check with Masoud.

jni2000 commented:
Kate, we have addressed your recent review comments. The last one you added 8 hours ago will be addressed in next revision. Please take a look. If there is no more concerns, we will start working on all related protection test cases. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -101,55 +101,50 @@ def generate_MCP_1_default_config(self, filename):
                                                                 pal_high_frequency,
                                                                 'test_user_1')
-    # Load DPA record
+    # Define DPA's 
```
No apostrophe.

jni2000 replied:
Updated   as 'DPAs'

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -101,55 +101,50 @@ def generate_MCP_1_default_config(self, filename):
+    protected_entities_iteration_0 = {
         'palRecords': pal_records_1,
-        'ppaRecords': ppa_records_1, 
+        'ppaRecords': ppa_record_1, 
```
This needs to be a list. Please be more careful.

jni2000 replied:
updated   as lists

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -219,37 +214,37 @@ def generate_MCP_1_default_config(self, filename):
 
     # Generate SAS Test Harnesses dump records for multiple iterations   
-    sas_test_harness_0_dump_records_iteration_0 = {
+    sas_test_harness_0_iteration_0_dump_records = {
```
I have asked several times for you to unify the convention: (iter #, item #) vs. (item #, iter #); you keep saying that it is "done" but the problem still exists. Please pick one convention and use it for all vars in this function.

jni2000 replied:
updated   followed iter#, item# through out the configurations

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,262 @@
+                                                                'test_user_1')
+    # Define DPA's 
+    dpa_1 = {
+       'dpaId': 'dpa_east_4',
```
I don't think this is a valid DPA ID/name. Please double-check the KML provided by NTIA/DoD.

jni2000 replied:
Ok, We can make them inline with what is currently used in other test cases like HB.12.

jni2000 commented:
Kate, we have addressed your recent comments except the following one, which we are working on right new and will update in next drop.

"I believe that the intention is to use individual CBSDs in every step where "DP test harness" is mentioned: so ITS can use N DPs and M individual CBSDs in each of these steps. The test spec is unfortunately not very clear on this point; you are welcome to double-check with Masoud."

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -140,8 +148,8 @@ def generate_MCP_1_default_config(self, filename):
     protected_entities_iteration_0 = {
-        'palRecords': pal_records_1,
-        'ppaRecords': ppa_record_1, 
+        'palRecords': [pal_records_1],
```
This should not be a list of a list. Please match with the documentation in our meeting minutes. It is a waste of our time to go around and around on these issues which are very clearly documented.

jni2000 replied:
Updated,   removed from list of list, now it’s a list

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -106,17 +114,17 @@ def generate_MCP_1_default_config(self, filename):
+    # Define DPAs
     dpa_1 = {
-       'dpaId': 'dpa_east_4',
+       'dpaId': 'east_dpa4',
```
Please use the correct DPA names.

jni2000 replied:
Updated   DPA ids as "east_dpa_4", "east_dpa_5" and   "east_dpa_6"

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -214,33 +222,37 @@ def generate_MCP_1_default_config(self, filename):
     iteration0_config = {
-        'cbsdRequests': [cbsd_records_iteration_0_domain_proxy_0,cbsd_records_iteration_0_domain_proxy_1],
+        'cbsdRequestsWithDomainProxys': [cbsd_records_iteration_0_domain_proxy_0,cbsd_records_iteration_0_domain_proxy_1],
+        'registrationRequests': [device_7],
```
Typically we take them as an array because that is easy to do with a DP. In the case of an individual CBSD, it is probably better to put the information for each CBSD into its own dict so that we can also specify the cert/key pair to be used.

jni2000 replied:
Updated   cbsd record as suggested in the comment

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -46,6 +46,10 @@ def generate_MCP_1_default_config(self, filename):
       open(os.path.join('testcases', 'testdata', 'device_f.json')))
+    device_7 = json.load(
+      open(os.path.join('testcases', 'testdata', 'device_g.json')))
+    device_8 = json.load(
```
This device is Cat B so you need to load conditional reg data.

jni2000 replied:
Updated   the conditional of device_8

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -214,33 +222,37 @@ def generate_MCP_1_default_config(self, filename):
     # Create the actual config.
     iteration0_config = {
-        'cbsdRequests': [cbsd_records_iteration_0_domain_proxy_0,cbsd_records_iteration_0_domain_proxy_1],
+        'cbsdRequestsWithDomainProxys': [cbsd_records_iteration_0_domain_proxy_0,cbsd_records_iteration_0_domain_proxy_1],
```
The correct plural of "proxy" is "proxies".

jni2000 replied:
Updated   , "proxys" to "proxies"

jni2000 commented:
Kate, we have responded to your review comments on mar. 15th. Please review and let us know if this good to pass the configuration portion of the MCP.1. Thanks.

jni2000 commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -237,18 +246,26 @@ def generate_MCP_1_default_config(self, filename):
-        'registrationRequests': [device_7],
-        'grantRequests': [grant_request_7],
+        'cbsdRequestsWithDomainProxies': [cbsd_records_iteration_0_domain_proxy_0,cbsd_records_iteration_0_domain_proxy_1],
+        'cbsdRecord' : {
```
Modified as list to support multiple individual CBSDs.

jni2000 commented:
Kate, we have addressed you latest review comment. Please take a look. We are expecting this configuration portion be finished up today to unblock all other xPR, EPR, PPR and GPR test cases. Please help us moving forward. Thanks.

jni2000 commented:
Kate and all other reviewers:

Can we please wrap up the configuration portion of this MCP.1 review by today to unblock the implementation of other derived incumbent protection test cases like xPRs, EPRs, FPRs, PPRs, GPRs and so on ? The review on this has gone through sufficient rounds and I believe we addressed all concerns. Please approve to allow moving forward if no more concerns.

Thanks.

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,295 @@
+        'airInterface': device_2['airInterface'],
+        'installationParam': device_2['installationParam'],
+        'measCapability': device_2['measCapability']
+    }
```
device_4 is also category B are we not preloading those conditionals?

jni2000 replied:
added for device 4.

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,295 @@
+    dpa_1 = {
+       'dpaId': 'east_dpa_4',
+       'frequencyRange': {'lowFrequency': 3650000000 , 'highFrequency': 3660000000 }
+    }
```
dpa freqs are wrong they should be 3550-3650

jni2000 replied:
changed.

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,295 @@
+    }
+    dpa_3 = {
+       'dpaId': 'east_dpa_6',
+       'frequencyRange': {'lowFrequency': 3670000000 , 'highFrequency': 3680000000 }
```
dpa freqs are wrong they should be 3550-3650

jni2000 replied:
changed

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,295 @@
+    dpa_2 = {
+       'dpaId': 'east_dpa_5',
+       'frequencyRange': {'lowFrequency': 3660000000 , 'highFrequency': 3670000000 }
+    }
```
dpa freqs are wrong they should be 3550-3650

jni2000 replied:
changed

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,295 @@
+
+    sas_test_harness_device_3 = json.load(
+      open(os.path.join('testcases', 'testdata', 'device_c.json')))
+    sas_test_harness_device_3['fccId'] = "test_fcc_id_i"
```
SUUT is only going to know fcc_ids a-h (the ones wthat will be (injected and registered) so the SA Test Harness is going to provide  FAD records that have FCC IDs that SUUT should considered in valid.....we've had the discussion internally about whether is the SAS's responsability to validate data on a FAD from a peer and while the protocol doesn't seem to say much about it it seems wrong for a SAS to accept CBSD data from a peer SAS while knowing that the FCCID soesn't exist

jni2000 replied:
What is you suggestion on this ? So far, looks like we will have to wait. But this will not affect the function as of this portion.

jni2000 replied:
Agreed. Do you have any suggestions on this ? As of new, our understanding is that this should work till we have any conclusions on what you mentioned and we can do adjustment by then.

jcereced commented on src/harness/testcases/testdata/dpa_0.json:
```diff
@@ -0,0 +1,36 @@
+        }
+   ]
+}
+
```
these dpa files should not be there.  please remove

jni2000 replied:
all test data are here. Where do you suggest them to be put in ?

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -247,11 +251,11 @@ def generate_MCP_1_default_config(self, filename):
-            'grantRequest': grant_request_7,
-            'clientCert': os.path.join('certs', 'client.cert'), 
-            'clientKey': os.path.join('certs', 'client.key')
+        'cbsdRecords' : {
```
Doesn't it make more sense to have a list of dictionaries instead of a dictionary of lists? You are much less likely to run into problems that way.

jni2000 replied:
we will study this and get back to you soon.

jni2000 replied:
Modified to list of  dictionaries to reduce ambiguity

jni2000 commented:
Kate and Javier

We have addressed your comments. Please continue and try to wrap up this to unblock the ones we need to move on. Thanks.

kate-harrison commented:
Since you are still studying my last (and only) comment, I don't think it's fair to ask me to re-review the code at this time. That said, the changes you made in response to Javi's comments look good (and thanks Javi for catching the bits I missed!).

jni2000 commented:
Kate, we have made changes to address your very last comment :), i.e. modified to list of dictionaries to reduce ambiguity. Please review. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -88,19 +97,23 @@ def generate_MCP_1_default_config(self, filename):
-    conditionals = {'registrationData': [conditionals_device_2, conditionals_device_8]} 
+    }
+
+    conditionals = {'registrationData': [conditionals_device_2, conditionals_device_4, conditionals_device_8]}
```
Depending on when PR #479 is submitted, this should probably be changed.

jni2000 replied:
thanks, we will wait on that approval and we will help review it.

jni2000 commented:
Thanks. We will wait for #479 approval and then do the merge.

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,308 @@
+
+    pal_low_frequency = 3550000000.0
+    pal_high_frequency = 3560000000.0
+
```
Please make frequencies integers

jni2000 replied:
fixed

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,308 @@
+    sas_test_harness_device_3['fccId'] = "test_fcc_id_i"
+    sas_test_harness_device_3['userId'] = "test_user_id_i"
+
+    sas_test_harness_device_4 = json.load(
```
still concerned about fcc_ids  on Sas Test Harness that SUUT does not know about, maybe we could change the cbsdSerialNumber instead and reuse known fcc_id?.  Will add an item on next week's meeting

jni2000 replied:
We can discuss this and make necessary changes if needed.

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,308 @@
+    # Load ESC record
+    esc_record_1 = json.load(
+      open(os.path.join('testcases', 'testdata', 'esc_sensor_record_0.json')))
+
```
esc_sensor_record_0.json is incomplete...missing protectionLevel (can be fixed outside this PR)

kate-harrison replied:
protectionLevel is optional, isn't it?

jcereced replied:
I guess it is......it would be so much simpler if things were just required or not..................

The protection level to be applied to this ESC sensor in units of dBm/MHz. If not present, the level should be interpreted as equal to that indicated in (n.8, R2-ESC-07)

jni2000 replied:
Yes, protectionLevel is optional. We may need to discuss if we need to add it in for the test case.

jcereced commented on src/harness/testcases/WINNF_FT_S_MCP_testcase.py:
```diff
@@ -0,0 +1,308 @@
+                                                                 pal_low_frequency,
+                                                                 pal_high_frequency,
+                                                                 'test_user_1')
+    # Define DPAs
```
the PPA is missing the CbsdReferenceId

jni2000 replied:
These are DPAs, are you referring to line 129 ? We will take a look of the PPA test data.

jcereced commented on src/harness/testcases/testdata/grant_1.json:
```diff
@@ -0,0 +1,9 @@
+    "operationFrequencyRange": {
+      "lowFrequency": 3630000000.0,
+      "highFrequency": 3640000000.0
+    }
```
please make frequencies integers

jni2000 replied:
fixed

jcereced commented on src/harness/testcases/testdata/grant_2.json:
```diff
@@ -0,0 +1,9 @@
+    "maxEirp": 10,
+    "operationFrequencyRange": {
+      "lowFrequency": 3650000000.0,
+      "highFrequency": 3660000000.0
```
please make frequencies integers

jni2000 replied:
fixed

jcereced commented on src/harness/testcases/testdata/grant_3.json:
```diff
@@ -0,0 +1,9 @@
+    "operationFrequencyRange": {
+      "lowFrequency": 366000000.0,
+      "highFrequency": 3670000000.0
+    }
```
please make frequencies integers

jni2000 replied:
fixed

kate-harrison commented on src/harness/sas_test_harness.cfg:
```diff
@@ -0,0 +1,14 @@
+[TestHarnessConfig]
```
I don't think this belongs in this PR.

kate-harrison commented on src/harness/sas_test_harness_interface.py:
```diff
@@ -0,0 +1,51 @@
+import abc
```
This is unrelated, please remove.

jni2000 replied:
removed
